src/Period.php

```diff
@@ -509,6 +530,37 @@ protected function roundDate(DateTimeInterface $date, int $precision): DateTimeI
         );
     }
 
+    protected static function convertDateIntervalToPrecision(DateInterval $interval): int
+    {
+        $now = new DateTimeImmutable();
+        $target = $now->add($interval);
+
+        if ($target == $now->add(new DateInterval('PT1S'))) {
```

Comment:
Can't we read the information out of `DateInterval` instead of doing these comparisons? Its properties are public: https://www.php.net/manual/en/class.dateinterval.php

Reply:
They are, but they are not reliable and have too many variations

```php
new DateInterval('P30D')); could have been intended as 1 Month
new DateInterval('PT1H'); // s = 0, h = 1
new DateInterval('PT60S'); // s = 60, h = 0

new DateInterval('P2Y13M64DT25H63M129S'); // I won't even break down this one :)
```

Reply:
What would the last example be mapped to though? 

Take a look at these properties on `DateInterval`

```php
public integer $y ;
public integer $m ;
public integer $d ;
public integer $h ;
public integer $i ;
public integer $s ;
```

Can't we say that if `$y !== 0` the precision should be years, etc? Something like this:

```php
$mapping = [
    'y' => Precision::YEAR,
    'm' => Precision::MONTH,
    // …
    's' => Precision::SECOND,
]

foreach ($mapping as $property => $precision) {
    if ($interval->${$property} !== 0) {
        return $precision;
    }
}
```

Reply:
Uh, nice! This would also remove the requirement that the source must have the exact interval ❤️. Will do!

Reply:
What with the case of a case like`PT1M15S`? Is the higher or lower measure the priority?

Reply:
Sounds like muted error. If you pass `P20D`, `->getIncludedDate()` will be date - 1 day when it should be date - 20D, it's even worth for composite intervals.

I think `Period` should actually throw an error if you pass an interval it does not handle, eg. anything with more than 1 of a unique base unit.

Note for the fun: `new DateInterval('P1W')` is valid and give you a 7-days interval.

Reply:
Our package doesn't support these complex intervals though, and I don't think we'll ever add support for them. So the question becomes whether to throw an exception or to infer the best interval.

I'd infer the interval for now. If there ever arise any problems because of this, we can reevaluate our solution.

Reply:
It's not "infer" if I give you 2 periods with week interval and ask `Period` if the periods `touchesWith` the result is just wrong because `Period` did not tell me week has been converted to day :/

Reply:
Asking for an exact match and throw an error for other intervals is the only thing to do. You're not guessing what the user meant, because he meant the interval he gave, and 24H may mean 1D (DST can imply 23/25H in 1 day).

Reply:
Good point. Let's throw an exception then

Reply:
`DateInterval` has no week property and will convert weeks to days internally

```php
$a = new DateInterval('P1W');
$b = new DateInterval('P7D');

assert($a->d === $b->d);
```

I think it would be okay to infer the precision and use the lowest denominator, and it will take into account the user's intent.

If they provide a period with an interval with "1D", the precision is "days", if they provide one with "24H", the precision is "hours", if they provide "1D12H", the precision is "hours" as well.

With that, we're not guessing what the user meant, we're inferring what he meant by looking at the specifics of the interval.

Reply:
The user obviously did not intent a day precision when he pass 7/20/100 day or even 1 day + 12 hours. My point is the user is not supposing to know the `Period` constraints and may just think you're able to use any interval as precision. Because it's not the case you should warn him instead of inferring. Inferring is right for "2" > 2 or 2.00001 > 2 but not 48H > 1H, this is far away from a legit inferring.

Reply:
I‘ve updated the PR with the map approach proposed by Brent. At the moment it doesn’t throw an exception, but the only case the default case would kick in is when a user provides an interval that only consists of `0` values.

Reply:
In my opinion, the user in this case is a developer, and when they use a library that has a “precision” concept and build a DatePeriod with an Interval to pass it over to Period, they think about how they’re building it.

But I’m fine with both approaches, it’s easy enough to change `!== 0` to `=== 1` and throw an exception if none match.

Reply:
Please note that `P1W` will still result in `Precision::DAY`. Introducing `Precision::WEEK` would need more changes and that’s for another day :)

Reply:
The early-return way using `=== 1` will also allow `P1M1DT1H` etc. which IMO should also throw an exception.